File: internal/types/builtins.go
package types

import "github.com/iley/pirx/internal/ast"

func getBuiltins() []FuncProto {
	return []FuncProto{
		{
			Name:       "printf",
			Args:       []Arg{{"fmt", ast.String}},
			ReturnType: ast.Int,
			Variadic:   true,
			External:   true,
		},
		{
			Name:       "putchar",
			Args:       []Arg{{"ch", ast.Int}},
			ReturnType: ast.Int,
			External:   true,
		},
		{
			Name:         "dispose",
			ExternalName: "Pirx_Dispose",
			Args:         []Arg{{"p", ast.Disposable}},
			External:     true,
		},
	}
}

File: ocaml/lib/typecheck/builtins.ml
module Type = Pirx_types.Type

let protos : Proto.t list =
  [
    {
      name = "PirxString";
      external_name = None;
      args = [ ("len", Type.Int); ("value", Type.Pointer Type.Int8) ];
      ret_type = Type.String;
      variadic = false;
      external_ = true;
    };
    {
      name = "printf";
      external_name = Some "PirxPrintf";
      args = [ ("fmt", Type.String) ];
      ret_type = Type.Void;
      variadic = true;
      external_ = true;
    };
    {
      name = "putchar";
      external_name = None;
      args = [ ("ch", Type.Int) ];
      ret_type = Type.Void;
      variadic = false;
      external_ = true;
    };
  ]
